x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskReplicationTests.java

```diff
@@ -221,6 +227,55 @@ public void testChangeFollowerHistoryUUID() throws Exception {
         }
     }
 
+    public void testRetryBulkShardOperations() throws Exception {
+        try (ReplicationGroup leaderGroup = createGroup(between(0, 1));
+             ReplicationGroup followerGroup = createFollowGroup(between(1, 3))) {
+            leaderGroup.startAll();
+            followerGroup.startAll();
+            leaderGroup.appendDocs(between(10, 100));
+            leaderGroup.refresh("test");
+            for (String deleteId : randomSubsetOf(IndexShardTestCase.getShardDocUIDs(leaderGroup.getPrimary()))) {
+                BulkItemResponse resp = leaderGroup.delete(new DeleteRequest("test", "type", deleteId));
+                assertThat(resp.getFailure(), nullValue());
+            }
+            leaderGroup.syncGlobalCheckpoint();
+            IndexShard leadingPrimary = leaderGroup.getPrimary();
+            // Simulates some bulk requests are completed on the primary and replicated to some (but all) replicas of the follower
+            // but the primary of the follower crashed before these requests completed.
+            for (int numBulks = between(1, 5), i = 0; i < numBulks; i++) {
+                long fromSeqNo = randomLongBetween(0, leadingPrimary.getGlobalCheckpoint());
+                long toSeqNo = randomLongBetween(fromSeqNo, leadingPrimary.getGlobalCheckpoint());
+                int numOps = Math.toIntExact(toSeqNo + 1 - fromSeqNo);
+                Translog.Operation[] ops = ShardChangesAction.getOperations(leadingPrimary, leadingPrimary.getGlobalCheckpoint(),
+                    fromSeqNo, numOps, leadingPrimary.getHistoryUUID(), new ByteSizeValue(Long.MAX_VALUE, ByteSizeUnit.BYTES));
+
+                IndexShard followingPrimary = followerGroup.getPrimary();
+                TransportWriteAction.WritePrimaryResult<BulkShardOperationsRequest, BulkShardOperationsResponse> primaryResult =
+                    TransportBulkShardOperationsAction.shardOperationOnPrimary(followingPrimary.shardId(),
+                        followingPrimary.getHistoryUUID(), Arrays.asList(ops),
+                        leadingPrimary.getMaxSeqNoOfUpdatesOrDeletes(), followingPrimary, logger);
+
+                IndexShard replica = randomFrom(followerGroup.getReplicas());
```

Comment:
maybe select a random subset ? then we will also test things that were fully replicated.